keps/sig-storage/20190129-csi-volume-resizing.md

```diff
@@ -0,0 +1,167 @@
+- if `ControllerExpandVolumeResponse` returns `false` in `node_expansion_required` then volume resize operation will be marked finished and both `pvc.Status.Capacity` and `pv.Spec.Capacity` will report updated value.
+
+If plugin does not implement `NodeExpandVolume` then volume resize operation will be marked as finished and both `pvc.Status.Capacity` and `pv.Spec.Capacity` will report updated value after successful completion of `ControllerExpandVolume` RPC call.
+
+If `ControllerExpandVolume`  call fails:
+- Then PVC will retain `Resizing` condition and will have appropriate events added to the PVC.
+- Controller will retry resizing operation with exponential backoff, assuming it corrects itself.
+
+A general mechanism for recovering from resize failure will be implemented via: https://github.com/kubernetes/kubernetes/issues/73036
+
+#### Expansion on Kubelet
+
+A CSI volume may require expansion on the node to finish volume resizing. In some cases - the entire resizing operation can happen on the node and
+plugin may choose to not implement `ControllerExpandVolume` CSI RPC call at all.
+
+Currently Kubernetes supports two modes of performing volume resize on kubelet. We will describe each mode here. For more information , please refer to original volume resize proposal - https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/grow-volume-size.md.
+
+
+##### Offline volume resizing on kubelet:
+
+This is the default mode and in this mode `NodeExpandVolume` will only be called when volume is being mounted on the node. In other words, pod that was using the volume must be re-created for expansion on node to happen.
+
+When a pod that is using the PVC is started, kubelet will compare `pvc.spec.resources.requests.storage` and `pvc.Status.Capacity`. It also compares PVC's size with `pv.Spec.Capacity` and if it detects PV is reporting same size as pvc's spec but PVC's status is still reporting smaller value then it determines -
+a volume expansion is pending on the node.  At this point if plugin implements `NodeExpandVolume` RPC call then, kubelet will call it and:
+
+If `NodeExpandVolume` is successful:
+- It will update `pvc.Status.Capacity` with latest value and remove all resizing related conditions from PVC.
+
+If `NodeExpandVolume` failed:
+- It will add a event to both PVC and Pod about failed resizing and resize operation will be retried. This will prevent pod from starting up.
+
+
+###### Online volume resizing on kubelet:
+
+More details about online resizing can be found in [Online resizing design](https://github.com/kubernetes/enhancements/pull/737) but essentially if
+`ExpandInUsePersistentVolumes` feature(currently Alpha) is enabled then kubelet will periodically poll all PVCs that are being used on the node and compare `pvc.spec.resources.requests.storage` and `pvc.Status.Capacity`(also `pv.Spec.Capacity`) and make similar determination about whether node expansion is required for the volume.
```

Comment:
remove "currently alpha". that will get out of date.  We can keep alpha/beta/ga status under "Implementation History"